File: frontend/src/app/core/services/favorite.service.ts
import { Injectable, signal, inject } from '@angular/core';
import { Observable, tap } from 'rxjs';
import { ApiService } from './api.service';

export interface FavoriteItem {
  id: string;
  product_id: string;
  collection_id?: string;
  price_at_save: number;
  product: any;
  collection?: { id: string; name: string };
}

export interface FavoriteCollection {
  id: string;
  name: string;
  is_public: boolean;
  items_count: number;
}

@Injectable({ providedIn: 'root' })
export class FavoriteService {
  private api = inject(ApiService);
  favorites = signal<FavoriteItem[]>([]);
  collections = signal<FavoriteCollection[]>([]);
  count = signal(0);

  getFavorites(): Observable<any> {
    return this.api.get<any>('favorites').pipe(
      tap(res => this.favorites.set(res.data || []))
    );
  }

  toggleFavorite(productId: string, collectionId?: string): Observable<any> {
    return this.api.post<any>('favorites/toggle', { product_id: productId, collection_id: collectionId }).pipe(
      tap(res => { if (res.favorites_count !== undefined) this.count.set(res.favorites_count); })
    );
  }

  getCollections(): Observable<any> {
    return this.api.get<any>('favorites/collections').pipe(
      tap(res => this.collections.set(res.collections || []))
    );
  }

  createCollection(name: string, isPublic = false): Observable<any> {
    return this.api.post<any>('favorites/collections', { name, is_public: isPublic });
  }

  getCount(): Observable<any> {
    return this.api.get<any>('favorites/count').pipe(tap(res => this.count.set(res.count)));
  }
}

File: flutter_app/lib/services/favorite_service.dart
import '../models/favorite.dart';
import 'api_service.dart';

class FavoriteService {
  final ApiService _api;

  FavoriteService(this._api);

  Future<List<FavoriteItem>> getFavorites({String? collectionId}) async {
    final params = <String, dynamic>{};
    if (collectionId != null) params['collection_id'] = collectionId;
    final response =
        await _api.get('/favorites', queryParameters: params);
    final list = response.data['favorites'] ?? response.data['data'] ?? [];
    return (list as List).map((e) => FavoriteItem.fromJson(e)).toList();
  }

  Future<Map<String, dynamic>> toggleFavorite(String productId,
      {String? collectionId}) async {
    final response = await _api.post('/favorites/toggle', data: {
      'product_id': productId,
      if (collectionId != null) 'collection_id': collectionId,
    });
    return response.data;
  }

  Future<List<FavoriteCollection>> getCollections() async {
    final response = await _api.get('/favorites/collections');
    final list =
        response.data['collections'] ?? response.data['data'] ?? [];
    return (list as List)
        .map((e) => FavoriteCollection.fromJson(e))
        .toList();
  }

  Future<FavoriteCollection> createCollection(String name) async {
    final response =
        await _api.post('/favorites/collections', data: {'name': name});
    return FavoriteCollection.fromJson(
        response.data['collection'] ?? response.data);
  }

  Future<int> getCount() async {
    final response = await _api.get('/favorites/count');
    return response.data['count'] ?? 0;
  }
}
